README.md
# voynich-ipsum
Generate non-latinate random text from a conlang.

## Summary

`voynich-ipsum` gives you a constructed language generator. You can specify various properties of the language -- it is synthetic
or analytic; it is VSO or SVO or OSV; it has a simple or complex phoneme inventoty; etc. -- or leave this up to chance. The language
is defined by the settings of various parameters and the construction of various probability distributions during instantiation. After
this it has a fixed name and properties. You can ask it to generate random text with our without a list of topics.

Mostly this is intended as a replacement for boring, repetitive lorem ipsum text.

## Why "voynich-ipsum"?

The [Voynich Manuscript](https://en.wikipedia.org/wiki/Voynich_manuscript) is a famous example of a text that is maybe real, maybe a fraud.
Examining the statistical properties of the characters, words, and structure of the text suggests mostly that is has the structure of a real
language, but not entirely, and any affiliation with an existing language and/or writing system is still up in the air.

The aim with `voynich-ipsum` is to generate explicitly fake languages with the same properties. The code incorporates many real linguistic
properties to produce a patina of plausibility for text that means nothing at all.

## Synopsis

```typescript
import { Language } from 'voynich-ipsum'

const lang = new Language({seed: 1})
console.log(lang.name)
// => Ekogo

// produce a 5-paragraph essay on a random topic
console.log(lang.essay(5))
// => 
// Ezozsebe azeghi ze. Ghoxeexl sedgog ghudlsooz. Ghodoghi aso akuoed do buoghupe.
//
// Saxeexl zaib adiupu.
//
// Uea ghuzazseu.
//
// Kakoko uvbnsu. Kakoko ghudili. Akoko zo e ghuipu be.
//
// Kakoko ghusodobabo. Sazope aakupo.

// assert something about 'foo'
console.log(lang.assertion('foo'))
// => Ufoo adiode.

// ask a question about 'foo'
console.log(lang.question('foo'))
// => Efoo ovde?

// shout something about 'foo'
console.log(lang.exclamation('foo'))
// => Foo doeeavuzl zeze!

// a phrase concerning foo
console.log(lang.nounPhrase('foo'))
// => afoo
```

Please look at the source code for more guidance. The public methods are documented.

## Caveats

The choices in here are based in some cases on published research in linguistic typology, sometimes in statistics I scraped out of
random text I found on the Internet, and most often out of plausible-seeming numbers I pulled out of my ass. There is a *vast* amount
of linguistic variation I have not covered. This does not produce ergative-absolutive languages (or nominative-accusative languages,
for that matter). It does not know about vowel harmony. It can't produce languages with tone, or nasal vowels. It can't do noun class
agreement between nouns and adjectives. It doesn't really know about adjectives at all. It produces phonemes that were easy to produce
with a mostly latinate character set. It has no variation in orthography. It never produces a right-to-left language. Etc. Etc. Etc.

I had great ambitions for covering typological variables and gave up on most of them because I'm lazy and my attention wanders.

### Sub-caveat

Because parameters are set by a random number generator during initialization, code changes are liable to completely change the language
generated with a particular random number seed. If this matters to you, you should specify a particular version number in your
dependencies so your languages won't change dramatically if a new version of lorem-voynich is released. If you're just happy to get
random gibberish, don't worry about this.

## Contributing

If you are less lazy or more knowledgeable than me, send me pull requests! Because I am lazy and my attention wanders, I may not respond to them promptly.
I apologize for that in advance.

## Acknowledgements

I would like to thank my wife and kids, who put up with too much of my nonsense and random enthusiasms.

I would also like to thank my employer, Green River, who also show much patience for my vagaries.

This was all done on my own time for reasons which are not clear even to me.


CHANGES.md
# Changelog

## v1.2.0
- switched to more efficient version of pick-me
## v1.1.1 *2022-8-6*
- fixed bug where `seed` parameter was being ignored by constructor
## v1.1.0 *2022-7-24*
- improved `config` method: more parameters are returned and the phonology config is no longer empty
## v1.0.1 *2022-7-23*
- fixed embarrassing copy-paste error
## v1.0.0 *2022-7-23*
- initial release


src/__tests__/utility.test.ts
import { capitalize, combinations, compact, subsets, titleize } from '../util'

describe('compact', () => {
  test('gets rid of both null and undefined', () => expect(compact([1, null, 2, undefined])).toEqual([1, 2]))
})

describe('combinations', () => {
  test('doc example works', () =>
    expect(
      combinations([
        [1, 2],
        [3, 4],
        [5, 6],
      ]),
    ).toEqual([
      [1, 3, 5],
      [1, 3, 6],
      [1, 4, 5],
      [1, 4, 6],
      [2, 3, 5],
      [2, 3, 6],
      [2, 4, 5],
      [2, 4, 6],
    ]))
})

describe('capitalize', () => {
  test('foo', () => expect(capitalize('foo')).toEqual('Foo'))
  test('foo bar', () => expect(capitalize('foo bar')).toEqual('Foo bar'))
  test('Foo', () => expect(capitalize('Foo')).toEqual('Foo'))
})

describe('titleize', () => {
  test('foo', () => expect(titleize('foo')).toEqual('Foo'))
  test('foo bar', () => expect(titleize('foo bar')).toEqual('Foo Bar'))
  test(' foo  bar ', () => expect(titleize(' foo  bar ')).toEqual(' Foo  Bar '))
  test('foo Bar', () => expect(titleize('foo Bar')).toEqual('Foo Bar'))
})

describe('subsets', () => {
  const set = [1, 2, 3]
  const ssets = subsets(set).sort((a, b) => {
    const l = a.length - b.length
    if (l) return l
    for (let i = 0; i < a.length; i++) {
      if (a[i] < b[i]) return -1
      if (b[i] < a[i]) return 1
    }
    return 0
  })
  test('[1, 2, 3', () => expect(ssets).toEqual([[1], [2], [3], [1, 2], [1, 3], [2, 3], [1, 2, 3]]))
})


src/__tests__/documentation.test.ts
/** tests for all the code shown in documentation */

// yes, this is shameful, but what do you expect for a gibberish generator?

describe('README', () => {
  describe('Synopsis', () => {
    test('stubbed', () => expect(true).toEqual(true))
  })//
})

describe('index.ts', () => {
  describe('something', () => {
    test('stubbed', () => expect(true).toEqual(true))
  })
})


src/__tests__/functionality.test.ts
import {Language} from '../language'

// we could use more tests, but hey! this is just a gibberish generator!

describe('language', () => {
  for (let i = 0; i < 10; i++) {
    const language = new Language({seed: i})
    // language.scramble()
    test('name', () => expect(language.name).toBeDefined())
    language.scramble()
    test('essay', () => expect(language.essay(5)).toBeDefined())
    test('assertion', () => expect(language.assertion('foo')).toBeDefined())
    test('question', () => expect(language.question('foo')).toBeDefined())
    test('exclamation', () => expect(language.exclamation('foo')).toBeDefined())
    test('nounPhrase', () => expect(language.nounPhrase('foo')).toMatch(/foo/i))
  }
})


src/util.ts
import { pickMe, Rng } from 'pick-me-too'

export class Hmm {
  public rng: Rng
  constructor(rng: Rng) {
    this.rng = rng
  }

  public n(): number {
    return this.rng()
  }

  public maybe(threshold: number = 0.5): boolean {
    return this.rng() <= threshold
  }

  public fromRange(low: number, high: number): number {
    if (low === high) return low
    if (low > high) throw new Error(`low (${low}) should be less than or equal to high (${high})`)
    return low + (high - low) * this.n()
  }
}

export function dupParams<T>(t: T): T {
  if (Array.isArray(t)) return t.map((o) => dupParams(o)) as any as T
  if (typeof t === 'object') {
    const rv: any = {}
    for (const [k, v] of Object.entries(t)) {
      rv[k] = dupParams(v)
    }
    return rv as T
  }
  return t
}

export function pickN<T>(n: number, picker: () => T): T[] {
  const seen: Set<T> = new Set()
  while (seen.size < n) seen.add(picker())
  return Array.from(seen)
}

// exhaustiveness checker
export function assertNever(x: never): never {
  console.error(x)
  throw new Error('Unexpected object: ' + x)
}

// for cloning simple structs
export function simpleClone<T>(t: T): T {
  switch (typeof t) {
    case 'undefined':
    case 'number':
    case 'boolean':
      return t
    case 'string':
      return ('' + t) as any as T
    default:
      if (null === t) return t
      if (Array.isArray(t)) {
        return t.map((v) => simpleClone(v)) as any as T
      }
      const obj: any = {}
      for (const [k, v] of Object.entries(t)) {
        obj[k] = simpleClone(v)
      }
      return obj as T
  }
}

// remove nullish things from an array
export function compact<T>(ar: (T | null | undefined)[]): T[] {
  const compacted: T[] = []
  for (const t of ar) {
    if (t != null) compacted.push(t)
  }
  return compacted
}

// turns [[1, 2], [3, 4], [5, 6]] into
// [[1, 3, 5], [1, 3, 6], [1, 4, 5], [1, 4, 6], [2, 3, 5], [2, 3, 6], [2, 4, 5], [2, 4, 6]]
export function combinations<T>(parts: T[][]): T[][] {
  if (parts.length === 1) return parts[0].map((v) => [v])
  const [head, ...tail] = parts
  const combos: T[][] = []
  const tailCombos = combinations(tail)
  for (const t of head) {
    for (const rest of tailCombos) {
      combos.push([t].concat(rest))
    }
  }
  return combos
}

// turns [1, 2, 3] into
// [[1], [2], [3], [1, 2], [1, 3], [2, 3], [1, 2, 3]] (not in this order)
export function subsets<T>(options: T[]): T[][] {
  if (options.length === 0) return []
  const [head, ...tail] = options
  const rv: T[][] = []
  const subsubset = subsets(tail)
  rv.push([head])
  for (const ss of subsubset) {
    rv.push(ss)
    rv.push([head, ...ss])
  }
  return rv
}

// mechanism to filter duplicates from a list
export function uniqBy<T, K>(ar: T[], test: (t: T) => K): T[] {
  const seen: Set<K> = new Set()
  const copy: T[] = []
  for (const t of ar) {
    const k = test(t)
    if (seen.has(k)) continue
    seen.add(k)
    copy.push(t)
  }
  return copy
}

// see https://stackoverflow.com/a/2450976
export function shuffle<T>(array: T[], rng: Rng) {
  let currentIndex = array.length,
    randomIndex
  while (currentIndex != 0) {
    randomIndex = Math.floor(rng() * currentIndex)
    currentIndex--
    [array[currentIndex], array[randomIndex]] = [array[randomIndex], array[currentIndex]]
  }
  return array
}


export function capitalize(s: string): string {
  return s.charAt(0).toUpperCase() + s.slice(1)
}

export function titleize(s: string): string {
  return s.split(/(\s+)/).map(s => capitalize(s)).join('')
}

src/language.ts
import { pickMe, pickMeToo, rando, Rng } from 'pick-me-too'
import { Morphology, MorphologyEngine } from './morphology'
import { Phonology, PhonologyEngine } from './phonology'
import { Syntax, SyntaxEngine } from './syntax'
import { Hmm, titleize } from './util'

export type LanguageParams = {
  seed?: number
  name?: string
  phonology?: Phonology
  morphology?: Morphology
  syntax?: Syntax
  rngGenerator?: (seed: number) => Rng
}

/**
 * A generator of gibberish.
 */
export class Language {
  /**
   * The language's phonology engine.
   */
  phonology: Readonly<PhonologyEngine>
  /**
   * The language's morophology engine.
   */
  morphology: Readonly<MorphologyEngine>
  /**
   * The language's syntactic engine.
   */
  syntax: Readonly<SyntaxEngine>
  /**
   * The name of the language in the language itself.
   */
  name: Readonly<string>
  private rng: Rng
  private sentenceCount: () => number
  private sentenceType: () => string
  private topicCount: () => number
  private subtopicCount: () => number
  private seed: number
  /**
   * Creates an instance of language.
   * @param [{ seed, name, phonology, morphology, syntax, rngGenerator }] 
   */
  constructor({ seed, name, phonology, morphology, syntax, rngGenerator }: LanguageParams = {}) {
    this.seed = seed ?? Math.random() * 1000
    this.rng = (rngGenerator ?? rando)(this.seed)
    this.phonology = new PhonologyEngine(phonology, this.rng)
    this.morphology = new MorphologyEngine(this.phonology, morphology, this.rng)
    this.syntax = new SyntaxEngine(this.morphology, syntax, this.rng)
    this.name = name ?? titleize(this.syntax.nounPhrase())
    this.sentenceCount = sentenceCount(this.rng)
    this.sentenceType = sentenceType(this.rng)
    this.topicCount = topicCount(this.rng)
    this.subtopicCount = subtopicCount(this.rng)
  }
  /**
   * Makes a complete sentence asserting a proposition, such as "My feet are cold.",
   * "So it is.", "Tomorrow and tomorrow and tomorrow creeps in this petty pace from day to day
   * to the last syllable of recorded time, and all our yesterdays have lighted fools the way to
   * dusty death.", and so forth.
   * 
   * @param topics - noun stems 
   * @returns assertion 
   */
  assertion(...topics: string[]): string {
    return this.syntax.assertion(...topics)
  }
  /**
   * Makes a complete sentence expression a question, such as "Do you want cheese?",
   * "Why me?", "Whose cousin are you again?", and so forth.
   * 
   * @param topics - noun stems 
   * @returns question 
   */
  question(...topics: string[]): string {
    return this.syntax.question(...topics)
  }
  /**
   * Makes a complete sentence expression an exclamation, such as "Begone, foul demon of the pit!",
   * "Not on my life!", "Bless your soul!", and so forth.
   * 
   * @param topics - noun stems 
   * @returns exclamation 
   */
  exclamation(...topics: string[]): string {
    return this.syntax.exclamation(...topics)
  }
  /**
   * Makes a random noun phrase, such as "the fat cat", "Ricardo", or "hemorrhagic fever".
   * 
   * @param [topic] - a noun stem
   * @returns phrase 
   */
  nounPhrase(topic?: string): string {
    return this.syntax.nounPhrase(topic)
  }
  private makeTopics(max?: number): string[] {
    const ar = []
    const seen = new Set()
    let n = this.topicCount()
    if (max && max < n) n = max
    while (ar.length < n) {
      const topic = this.morphology.nounStem()
      if (!seen.has(topic)) {
        seen.add(topic)
        ar.push(topic)
      }
    }
    return ar
  }
  /**
   * Makes a collection of sentences on a consistent set of topics.
   * @param [topics] - a set of noun stems
   * @returns paragraph 
   */
  paragraph(topics?: string[]): string {
    topics ??= this.makeTopics(4)
    let lim = 2 // the maximum number of topics per sentence
    if (lim > topics.length) lim = topics.length
    const frequencies: [string | undefined, number][] = []
    for (const t of topics) frequencies.push([t, this.rng()])
    let sum = 0
    for (const [, n] of frequencies) sum += n
    frequencies.push([undefined, sum * 2]) // only every third topic is an existing topic
    const topicPicker = pickMe(frequencies, this.rng)
    const sentenceCount = this.sentenceCount()
    const sentences: string[] = []
    for (let i = 0; i < sentenceCount; i++) {
      topics = []
      for (let j = 0; j < lim; j++) {
        const t = topicPicker()
        if (t && t !== topics[0]) topics.push(t)
      }
      switch (this.sentenceType()) {
        case 'assertion':
          sentences.push(this.assertion(...topics))
          break
        case 'question':
          sentences.push(this.question(...topics))
          break
        case 'exclamation':
          sentences.push(this.exclamation(...topics))
          break
      }
    }
    return sentences.join(' ')
  }
  /**
   * Makes a few paragraphs of text on a consistent set of topics.
   *
   * @param [paragraphs] - the number of paragraphs sought; if no number is provided, a random number between 3 and 15 is chosen
   * @param [topics] - a set of noun stems
   * @returns essay
   */
  essay(paragraphs?: number, topics?: string[]): string {
    paragraphs ??= Math.round(new Hmm(this.rng).fromRange(3, 15))
    if (paragraphs < 1) throw 'an essay must have a positive number of paragraphs'
    topics ??= this.makeTopics()
    const frequencies: [string, number][] = []
    for (const t of topics) frequencies.push([t, this.rng()])
    const topicPicker = pickMe(frequencies, this.rng)
    const text: string[] = []
    for (let j = 0; j < paragraphs; j++) {
      const subtopics: string[] = []
      const seen = new Set()
      let lim = this.subtopicCount()
      if (lim > topics.length) lim = topics.length
      for (let k = 0; k < lim; k++) {
        const t = topicPicker()
        if (!seen.has(t)) {
          seen.add(t)
          subtopics.push(t)
        }
      }
      text.push(this.paragraph(subtopics))
    }
    return text.join('\n\n')
  }
  /**
   * Randomly resets the seed of the random number generator to make
   * the next text produced unpredictable.
   */
  scramble() {
    ;(this.rng as any)(Math.random())
  }
  /**
   * Produces the configurable parameters of this language aside from the random number generator.
   *
   * If you find a language with a syntax that suits you, or partially suits you,
   * you can use this to obtain the configuration, tinker with it, and apply it to other languages.
   *
   * Note, the configuration should be regarded as read-only. In any case, if you modify the configuration
   * it will have no effect on the engine. The configuration parameters are only used during initialization.
   *
   * @returns language configuration
   */
  config(): Readonly<LanguageParams> {
    return {
      name: this.name,
      seed: this.seed,
      phonology: this.phonology.config(),
      morphology: this.morphology.config(),
      syntax: this.syntax.config(),
    }
  }
}

/*
to the extent possible we do the expensive part of working with probability distributions
at load time
*/

// frequencies extracted from a few NYT articles
const sentenceCount = pickMeToo([
  [1, 36],
  [2, 60],
  [3, 22],
  [4, 10],
  [5, 4],
  [6, 1], // added
  [7, 1], // added
  [8, 1], // added
  [9, 1],
  [10, 1], // added
  [11, 1],
])

// from the same articles
const sentenceType = pickMeToo([
  ['assertion', 291],
  ['exclamation', 4],
  ['question', 7],
])

// topics per essay
const topicCount = pickMeToo([
  // arbitrary
  [1, 15],
  [2, 20],
  [3, 35],
  [4, 40],
  [5, 30],
  [6, 20],
  [7, 10],
  [8, 5],
  [9, 3],
  [10, 2],
  [11, 1],
  [12, 1],
])

// topics per paragraph
const subtopicCount = pickMeToo([
  // arbitrary
  [1, 1],
  [2, 4],
  [3, 8],
  [4, 10],
])


src/morphology.ts
import { pickMe, pickMeToo, Rng } from 'pick-me-too'
import { Phonology, PhonologyEngine } from './phonology'
import { assertNever, Hmm, shuffle, uniqBy } from './util'

export type Morphology = {
  analytic?: boolean
  verbalInflections?: boolean
  nominalInflections?: boolean
  derivationalAdfixes?: boolean
  stemComplexity?: StemComplexity
  adfixStyle?: AdfixStyle
}

// where do inflectional suffixes go?
type AdfixStyle = 'prefix' | 'suffix' | 'both'

const adfixStyle = pickMeToo<AdfixStyle>([
  ['prefix', 3],
  ['suffix', 10],
  ['both', 1],
])

type StemComplexity = 'simple' | 'moderate' | 'complex'

const stemComplexity = pickMeToo<StemComplexity>([
  ['simple', 2],
  ['moderate', 4],
  ['complex', 1],
])

/**
 * A thing for building words.
 * 
 * A vast amount of morphology is ignored by this. There is no reduplication, for example.
 */
export class MorphologyEngine {
  private morphology: Morphology
  private phonologyEngine: Readonly<PhonologyEngine>
  private rng: Rng
  private stem!: () => string
  private phonology: Phonology
  private closedClassStems: Set<string>
  private adfixMaker!: () => string
  /**
   * Makes a small adverbish word. This will be from a closed class.
   */
  particle!: () => string
  /**
   * Makes a word with noun morphology. There is a closed class of common nouns
   * and an open class of less common nouns.
   */
  noun!: (stem?: string) => string
  /**
   * Picks a form from the closed class of pronouns.
   */
  pronoun!: () => string
  /**
   * Generates an adverbish thing.
   */
  adverb!: () => string
  /**
   * Makes a word with verbal morphology. There is a closed class of common verbs
   * and an open class of less common verbs.
   */
  verb!: (stem?: string) => string
  /**
   * Generates a noun stem. This may include derivational morphology but will not include
   * inflectional morphology.
   */
  nounStem!: () => string
  /**
   * Generates a verb stem. This may include derivational morphology but will not include
   * inflectional morphology.
   */
  verbStem!: () => string
  /**
   * Creates an instance of morphology engine.
   * @param p - the thing that will give us syllables
   * @param [m] - optional configuration parameters
   * @param [rng] - a random number generator with which to pick configuration parameters and then to build random words
   */
  constructor(p: Readonly<PhonologyEngine>, m: Morphology = {}, rng: Rng = () => Math.random()) {
    this.morphology = m
    this.phonologyEngine = p
    this.phonology = p.config()
    this.rng = rng
    this.closedClassStems = new Set()
    const h = new Hmm(rng)
    m.analytic ??= h.maybe(0.4)
    if (m.analytic) {
      m.verbalInflections = false
      m.nominalInflections = false
    } else {
      m.verbalInflections ??= h.maybe(0.9)
      m.nominalInflections ??= !m.verbalInflections || h.maybe(0.3)
    }
    m.derivationalAdfixes = !m.analytic || h.maybe(0.5)
    this.initializeAdfixMaker()
    this.initializeStemmer()
    this.initializeVerbs(h)
    this.initializeNominals(h)
    this.initializeAdverbs(h)
    this.particle = this.makeParticles(h.fromRange(10, 100), false)
  }

  // "adverbs" in this case just means not inflected
  private initializeAdverbs(h: Hmm) {
    const frequentAdverbLimit = h.fromRange(10, 50)
    const frequentAdverbProbability = h.fromRange(0.1, 0.5)
    const frequentAdverbs = []
    while (frequentAdverbs.length < frequentAdverbLimit) {
      const s = this.stem()
      if (this.addClosedClassItem(s)) frequentAdverbs.push(s)
    }
    const frequentFrequencies: [string, number][] = frequentAdverbs.map((s) => [s, h.n()])
    const frequentAdverbPicker = pickMe(frequentFrequencies, this.rng)
    this.adverb = () => (h.maybe(frequentAdverbProbability) ? frequentAdverbPicker() : this.stem())
  }

  // returns whether the item does not yet belong to the closed class
  // words in the language, adding it to the closed class if not
  private addClosedClassItem(s: string): boolean {
    if (!this.closedClassStems.has(s)) {
      this.closedClassStems.add(s)
      return true
    }
    return false
  }
  // create inflected noun and pronoun generators
  private initializeNominals(h: Hmm) {
    const lim = pickMe(
      [
        [3, 1],
        [4, 10],
        [5, 50],
        [6, 100],
        [7, 90],
        [8, 80],
        [9, 30],
        [10, 10],
        [11, 1],
      ],
      this.rng,
    )()
    const cc: string[] = []
    while (cc.length < lim) {
      const s = this.stem()
      if (this.addClosedClassItem(s)) {
        cc.push(s)
      }
    }
    const ccFrequencies = cc.map((s) => [s, h.n()] as [string, number])
    const inflector = this.makeInflector(h, false)
    const pronounPicker = pickMe(ccFrequencies, this.rng)
    this.pronoun = () => inflector(pronounPicker())
    const stem = this.makeInflectionalStem(h)
    this.nounStem = stem
    this.noun = (s?: string) => inflector(s ?? stem())
  }

  private makeInflector(h: Hmm, isVerb: boolean): (stem: string) => string {
    if (this.morphology.analytic) return (stem: string) => stem
    if (isVerb) {
      // potentially more complicated inflections
      let slotCount = pickMe(
        [
          [1, 5],
          [2, 3],
          [3, 2],
          [4, 1],
        ],
        this.rng,
      )()
      if (slotCount === 1 && this.morphology.adfixStyle === 'both') slotCount = 2
      const slots: (() => string)[] = []
      for (let i = 0; i < slotCount; i++) {
        slots.push(this.makeParticles(h.fromRange(2, 20), true))
      }
      let stemPosition = 0
      switch (this.morphology.adfixStyle!) {
        case 'prefix':
          stemPosition = slots.length
          break
        case 'suffix':
          stemPosition = 0
          break
        case 'both':
          stemPosition = Math.round(h.fromRange(1, slotCount - 1))
          break
        default:
          assertNever(this.morphology.adfixStyle! as never)
      }
      const lim = slots.length + 1
      return (stem: string) => {
        let s = ''
        for (let i = 0, j = 0; i < lim; i++) {
          s += i === stemPosition ? stem : slots[j++]()
        }
        return s
      }
    } else {
      // simpler inflections -- just one or two slots
      const slotCount = pickMe(
        [
          [1, 5],
          [2, 3],
        ],
        this.rng,
      )()
      let slots: (() => string)[] = []
      for (let i = 0; i < slotCount; i++) {
        const formCount =
          slots.length == 0
            ? pickMe(
                [
                  [4, 12],
                  [5, 11],
                  [6, 10],
                  [7, 9],
                  [8, 8],
                  [9, 7],
                  [10, 6],
                  [11, 4],
                  [12, 2],
                  [13, 1],
                  [14, 1],
                ],
                this.rng,
              )()
            : pickMe(
                // fewer this time
                [
                  [1, 5],
                  [2, 4],
                  [3, 2],
                  [4, 1],
                ],
                this.rng,
              )()
        slots.push(this.makeParticles(formCount, true))
      }
      // maybe make the shorter slot first
      if (slots.length === 2 && h.maybe(0.5)) {
        const [s1, s2] = slots
        slots = [s2, s1]
      }
      switch (this.morphology.adfixStyle!) {
        case 'prefix': {
          if (slots.length === 2) {
            const [s1, s2] = slots
            return (stem: string) => s1() + s2() + stem
          }
          const [s] = slots
          return (stem: string) => s() + stem
        }
        case 'suffix': {
          if (slots.length === 2) {
            const [s1, s2] = slots
            return (stem: string) => stem + s1() + s2()
          }
          const [s] = slots
          return (stem: string) => stem + s()
        }
        case 'both': {
          if (slots.length === 2) {
            const [s1, s2] = slots
            return (stem: string) => s1() + stem + s2()
          }
          const [s] = slots
          return (stem: string) => stem + s()
        }
        default:
          assertNever(this.morphology.adfixStyle! as never)
      }
    }
  }

  private makeInflectionalStem(h: Hmm) {
    const lim = h.fromRange(10, 100)
    const cc: string[] = []
    while (cc.length < lim) {
      const s = this.stem()
      if (this.addClosedClassItem(s)) {
        cc.push(s)
      }
    }
    const ccFrequencies = cc.map((s) => [s, h.n()] as [string, number])
    const ccStemPicker = pickMe(ccFrequencies, this.rng)
    let derivedStem = this.nonClosedClassStem.bind(this)
    if (this.morphology.derivationalAdfixes) {
      const adfixMakers: (() => string)[] = []
      for (let i = 0; i < 6; i++) adfixMakers.push(this.makeDerivationalAdfixes(h))
      const adfixCount = pickMe(
        [
          [0, 50],
          [1, 5],
          [2, 2],
          [3, 1],
        ],
        this.rng,
      )
      derivedStem = () => {
        const prefixCount = adfixCount(),
          suffixCount = adfixCount()
        if (prefixCount === 0 && suffixCount === 0) return this.nonClosedClassStem()
        let prefix = ''
        if (prefixCount > 0) {
          const ar = shuffle([0, 1, 2], this.rng).slice(0, prefixCount)
          ar.sort()
          for (const i of ar) {
            prefix += adfixMakers[i]()
          }
        }
        let suffix = ''
        if (suffixCount > 0) {
          const ar = shuffle([3, 4, 5], this.rng).slice(0, suffixCount)
          ar.sort()
          for (const i of ar) {
            suffix += adfixMakers[i]()
          }
        }
        return prefix + this.nonClosedClassStem() + suffix
      }
    }
    const threshold = h.fromRange(0.2, 0.35)
    return () => (h.maybe(threshold) ? ccStemPicker() : derivedStem())
  }

  private initializeAdfixMaker() {
    this.morphology.adfixStyle ??= adfixStyle(this.rng)()
    const syllableComplexity = this.phonology.numberPossibleSyllables ?? 0
    // the simpler the syllables the more syllables in the adfixes
    const frequencies: [number, number][] =
      syllableComplexity < 100
        ? [
            [1, 4],
            [2, 3],
            [3, 2],
            [4, 1],
          ]
        : syllableComplexity < 1000
        ? [
            [1, 5],
            [2, 2],
            [3, 1],
          ]
        : [
            [1, 5],
            [2, 1],
          ]
    const lengthPicker = pickMe(frequencies, this.rng)
    this.adfixMaker = () => {
      let s = ''
      // maximum of 1000 tries to find a novel adfix
      for (let attempt = 0; attempt < 1000; attempt++) {
        s = ''
        for (let i = 0, lim = lengthPicker(); i < lim; i++) s += this.phonologyEngine.syllable()
        if (this.addClosedClassItem(s)) return s
      }
      return s
    }
  }

  private nonClosedClassStem(): string {
    let stem = this.stem()
    while (this.closedClassStems.has(stem)) stem = this.stem()
    return stem
  }

  // generate a function that makes an uninflected stem
  private initializeStemmer() {
    this.morphology.stemComplexity ??= stemComplexity(this.rng)()
    let lengthPicker: () => number
    switch (this.morphology.stemComplexity) {
      case 'simple':
        lengthPicker = pickMe(
          [
            [1, 10],
            [2, 1],
          ],
          this.rng,
        )
        break
      case 'moderate':
        lengthPicker = pickMe(
          [
            [1, 10],
            [2, 5],
            [3, 1],
          ],
          this.rng,
        )
        break
      case 'complex':
        lengthPicker = pickMe(
          [
            [1, 10],
            [2, 10],
            [3, 5],
            [4, 2],
            [5, 1],
          ],
          this.rng,
        )
        break
      default:
        assertNever(this.morphology.stemComplexity)
    }
    this.stem = () => {
      let s = ''
      for (let i = 0, lim = lengthPicker(); i < lim; i++) {
        s += this.phonologyEngine.syllable()
      }
      return s
    }
  }
  private initializeVerbs(h: Hmm) {
    const stem = this.makeInflectionalStem(h)
    this.verbStem = stem
    const inflector = this.makeInflector(h, true)
    this.verb = (s?: string) => inflector(s ?? stem())
  }
  // useful for making inflections and particles
  private simpleSyllable(): string {
    if (this.phonology.closedSyllables || this.phonology.consonantClusters?.initial) {
      const ar: string[] = []
      // yeah, not so efficient, but we just use this during language initialization
      for (let i = 0; i < 4; i++) {
        ar.push(this.phonologyEngine.syllable())
      }
      ar.sort((a, b) => a.length - b.length)
      return ar[0]
    } else {
      return this.phonologyEngine.syllable()
    }
  }

  // returns a copy of the phonology (copy to prevent mistaken monkeybusiness)
  /**
   * Provides the configuration parameters used by the engine.
   * 
   * This configuration should be read-only. In any case, changing it will have no effect on
   * the behavior of the engine, since configuration parameters are only consulted during initialization.
   * 
   * If you find a language whose word shapes you like, you can use this to obtain its configuration and
   * then copy it to other languages.
   * 
   * @returns configuration parameters 
   */
  config(): Readonly<Morphology> {
    return this.morphology
  }

  private makeDerivationalAdfixes(h: Hmm): () => string {
    const adfixMaker = this.initializeAdfixMaker()
    const adfixFrequencies: [string, number][] = []
    for (let i = 0, lim = h.fromRange(10, 15); i < lim; i++) {
      adfixFrequencies.push([this.adfixMaker(), h.n()])
    }
    return pickMe(
      uniqBy(adfixFrequencies, ([v]) => v),
      this.rng,
    )
  }

  /**
   * Creates a closed class of morphemes that will appear in a particular morphological or syntactic "slot".
   * This is a public method mostly because syntax engines need it.
   * 
   * @param n -- a number of "particles" to make
   * @param includeBlank -- whether the null particle should be among the particles
   * @returns a collection of morphemes
   */
  makeParticles(n: number, includeBlank: boolean): () => string {
    const seen: Set<string> = new Set()
    let found: string[] = []
    // NOTE: there is a possibility for an infinite loop here if we ask for too many variants!
    let safety = 0 // we'll just cut things off after 1000 samples if we get that far
    // we assume inflectional morphemes tend to have relatively simple syllable structure,
    // so for languages with complex syllable structure, we overgenerate candidates and filter
    // NOTE: to simplify things we aren't generating any multi-syllabic morphemes like Latin -ōrum

    while (found.length < n && safety < 1000) {
      const s = this.simpleSyllable()
      if (!seen.has(s)) {
        seen.add(s)
        found.push(s)
      }
      safety++
    }
    const frequencies = found.map((s) => [s, this.rng()] as [string, number])
    if (includeBlank) frequencies.push(['', this.rng() * 2]) // blanks are more common
    return pickMe(frequencies, this.rng)
  }
}
